### PR TITLE
FIX: missing post and topic edited webhooks

### PR DESCRIPTION
**Pull request by fantasticfears**

### CODE REVIEW

**Comment by ZogStriP** [comment]
No test?

**Comment by fantasticfears** [comment]
Sorry, fixed!

File: lib/post_revisor.rb [hunk 1]
```diff
@@ -405,6 +405,7 @@ def is_last_post?
   def plugin_callbacks
     DiscourseEvent.trigger(:before_edit_post, @post)
     DiscourseEvent.trigger(:validate_post, @post)
+    DiscourseEvent.trigger(:before_edit_topic, @topic, @editor) if topic_changed?
```
Comment (tgxworld) [inline lib/post_revisor.rb]: I think `:before_edit_post` is sufficient here and we don't need another event trigger. We can easily check if it is a topic by checking if the `post_number` is one.

Comment (tgxworld) [inline lib/post_revisor.rb]: I think `:before_edit_post` is sufficient here and we don't need another event trigger. We can easily check if it is a topic by checking if the `post_number` is one.

Comment (fantasticfears) [inline lib/post_revisor.rb]: Somehow it's needed to limit the scope of webhook, `post_number` is not only one condition, but also a `topic_changed?`. I'll need some way to check whether topic is changed. But `PostRevisor` is a huge object, I don't see any spots sending the whole object. So I would only send the single value in the event. Though it looks a little bit suspicious comparing others.

File: app/models/web_hook.rb [hunk 2]
```diff
@@ -57,18 +61,16 @@ def self.enqueue_topic_hooks(event, topic, user)
...
-        user_id: user&.id,
-        category_id: post.topic&.category&.id,
-        event_name: event.to_s
-      )
+      WebHook.enqueue_post_hooks(event, post, user)
     end
   end
 
+  DiscourseEvent.on(:before_edit_post) do |post, topic_changed|
+    WebHook.enqueue_post_hooks(:before_edit_post, post)
```
Comment (tgxworld) [inline app/models/web_hook.rb]: hmm shouldn't the event name be like `post edited` or `topic edited`?

Comment (fantasticfears) [inline app/models/web_hook.rb]: I was worried they may interested in where the event initially happens. But it seems they just need to distinguish them..so goes for `x_edited`

File: app/models/web_hook.rb [hunk 3]
```diff
@@ -57,18 +61,16 @@ def self.enqueue_topic_hooks(event, topic, user)
...
-        topic_id: post&.topic&.id,
-        user_id: user&.id,
-        category_id: post.topic&.category&.id,
-        event_name: event.to_s
-      )
+      WebHook.enqueue_post_hooks(event, post, user)
     end
   end
 
+  DiscourseEvent.on(:before_edit_post) do |post, topic_changed|
```
Comment (tgxworld) [inline app/models/web_hook.rb]: Hmm sorry I just realized that this event is called before the transaction has been committed. We probably want to add event triggers for after the post has been committed to the DB

**Comment by fantasticfears** [comment]
New DiscourseEvent `post_edited` added.

File: app/models/web_hook.rb [hunk 4]
```diff
@@ -40,10 +40,14 @@ def self.enqueue_hooks(type, opts = {})
     end
   end
 
-  def self.enqueue_topic_hooks(event, topic, user)
+  def self.enqueue_topic_hooks(event, topic, user=nil)
     WebHook.enqueue_hooks(:topic, topic_id: topic.id, user_id: user&.id, category_id: topic&.category&.id, event_name: event.to_s)
   end
 
+  def self.enqueue_post_hooks(event, post, user=nil)
+    WebHook.enqueue_hooks(:post, post_id: post.id, topic_id: post&.topic&.id, user_id: user&.id, category_id: post&.topic&.category&.id, event_name: event.to_s)
```
Comment (tgxworld) [inline app/models/web_hook.rb]: `post&.topic&.category&.id` similarily this can be `post&.topic.category_id`

**Pull request merged**